[KOGITO-8645] Review Conditions Status API

ricardozanini opened the pull request:
**Description of the change:**
In this PR, we change the way we deal with conditions following the Kubernetes convention.


**Motivation for the change:**
See: https://issues.redhat.com/browse/KOGITO-8645


**Checklist**

- [x] Add or Modify a unit test for your change
- [x] Have you verified that tall the tests are passing?

<details>
<summary>
How to backport a pull request to a different branch?
</summary>

In order to automatically create a **backporting pull request** please add one or more labels having the following format `backport-<branch-name>`, where `<branch-name>` is the name of the branch where the pull request must be backported to (e.g., `backport-7.67.x` to backport the original PR to the `7.67.x` branch).

> **NOTE**: **backporting** is an action aiming to move a change (usually a commit) from a branch (usually the main one) to another one, which is generally referring to a still maintained release branch. Keeping it simple: it is about to move a specific change or a set of them from one branch to another.

Once the original pull request is successfully merged, the automated action will create one backporting pull request per each label (with the previous format) that has been added.

If something goes wrong, the author will be notified and at this point a manual backporting is needed.

> **NOTE**: this automated backporting is triggered whenever a pull request on `main` branch is labeled or closed, but both conditions must be satisfied to get the new PR created.
</details>

ricardozanini commented:
@radtriste @davidesalerno @spolti feel free to review.

radtriste commented on config/manager/manager.yaml:
```diff
@@ -34,6 +34,8 @@ spec:
         image: controller:latest
         name: manager
         securityContext:
+          seccompProfile:
```
that one will create issues with older kubernetes ...

ricardozanini replied:
Yes, we need to update/create the docs with a minimum version in mind. This was causing problems for the e2e. We can change the tests, but I think we can afford to have this strict security set. wdyt?

radtriste replied:
we would still break with OCP EUS :S

radtriste commented on controllers/kogitoserverlessworkflow_controller.go:
```diff
@@ -29,9 +29,11 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 
+	"github.com/kiegroup/kogito-serverless-operator/api"
```
operatorapi ?

ricardozanini replied:
The operatorapi is this one:

```
operatorapi "github.com/kiegroup/kogito-serverless-operator/api/v1alpha08"
```

It's the versioned one where we can change only the import and keep the alias.

In this case, `api` is the better name.

radtriste reviewed (commented):
Why not reusing directly https://github.com/kubernetes/apimachinery/blob/release-1.23/pkg/apis/meta/v1/types.go#L1432-L1492 ?

ricardozanini commented:
> Why not reusing directly https://github.com/kubernetes/apimachinery/blob/release-1.23/pkg/apis/meta/v1/types.go#L1432-L1492 ?

 Because I want the liberty to change it and not be impacted if they change their API.

ricardozanini commented on api/condition_types.go:
```diff
@@ -0,0 +1,102 @@
+	// +required
+	Status v1.ConditionStatus `json:"status"`
+	// The last time this condition was updated.
+	LastUpdateTime metav1.Time `json:"lastUpdateTime,omitempty"`
```
It's not a transition, actually. It's an update. 😅

radtriste replied:
I was referring to the meta/v1 api :)

ricardozanini replied:
Oh ok! That's because we don't transition to the conditions, they are evaluated together every time. We could do it, but I don't see a benefit of doing it at the moment.

For instance, we have a `Built` condition that will be either `Unknow` or `True` in dev. We won't transition to it, since it's a static condition in the dev mode. The `LastUpdateTime` just clarifies that in both cases.

radtriste commented on api/status_types.go:
```diff
@@ -0,0 +1,345 @@
+	Conditions Conditions `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
+	// The generation observed by the deployment controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
```
see the meta/v1 spec. That `ObservedGeneration` is for each condition AFAIU

ricardozanini replied:
Oh sure, depends on the implementation. On Knative they use it in the whole status since every reconciliation cycle they reevaluate every condition. We are following a similar approach. So it makes sense keeping in a higher level.

Also in the Deployment Status, same thing:

```yaml
status:
  observedGeneration: 3
  replicas: 1
  updatedReplicas: 1
  readyReplicas: 1
  availableReplicas: 1
  conditions:
    - type: Available
      status: 'True'
      lastUpdateTime: '2023-02-28T11:25:36Z'
      lastTransitionTime: '2023-02-28T11:25:36Z'
      reason: MinimumReplicasAvailable
      message: Deployment has minimum availability.
    - type: Progressing
      status: 'True'
      lastUpdateTime: '2023-02-28T11:55:42Z'
      lastTransitionTime: '2023-02-28T11:25:32Z'
      reason: NewReplicaSetAvailable
      message: ReplicaSet "greeting-57f9548ff4" has successfully progressed.
```

radtriste commented on config/crd/bases/sw.kogito.kie.org_kogitoserverlessworkflows.yaml:
```diff
@@ -28,7 +28,7 @@ spec:
       name: Address
       type: string
-    - jsonPath: .status.condition
+    - jsonPath: .status.conditions[?(@.type=='Running')].status
```
Could we instead give the latest condition type ?
Or call it `RUNNING` instead of Condition ?

Also, I thin the `REASON` needs to be changed
```
$ kget KogitoServerlessWorkflow
NAME       PROFILE   VERSION   ADDRESS   CONDITION   REASON
greeting             0.0.1               False       
```

ricardozanini replied:
> Could we instead give the latest condition type ?

What do you mean?

Actually, I need to change this. It should be:

```
$ kget KogitoServerlessWorkflow
NAME       PROFILE   VERSION   ADDRESS   READY   REASON
greeting             0.0.1               False       
```

Condition > Ready. And the reason should be changed, as you pointed! Nice catch.

radtriste commented:
When testing, I tried with the `config/samples/sw.kogito_v1alpha08_kogitoserverlessworkflow.yaml` sample.
It gets built but the built image is launched as a pod.

And I think that is also the problem reported by the e2e tests.

Seems to be a reconciler issue or ?

ricardozanini commented:
>It gets built but the built image is launched as a pod.

wdym? I'll test it locally. This entire reconciliation must be reviewed, actually. I'll open a JIRA.

> And I think that is also the problem reported by the e2e tests.

hmm it might be related to the timing. I'll take a look at the new snapshot image. Might be a problem there.

ricardozanini commented:
> And I think that is also the problem reported by the e2e tests.

It was failing because we removed the `seccompProfile`. I've removed the test and the namespace with the labels enforcing it.

ricardozanini commented:
> It gets built but the built image is launched as a pod.

I'll review this use case tomorrow.

radtriste commented:
gonna have a new run this afternoon

ricardozanini commented:
The prod profile is still failing. It builds but does nothing. I'll fix it tomorrow.

radtriste commented:
@

> The prod profile is still failing. It builds but does nothing. I'll fix it tomorrow.

It could be related to the issue on runtimes ...

ricardozanini commented:
@radtriste all done.

desmax74 commented on builder/config.go:
```diff
@@ -46,23 +46,22 @@ func NewCustomConfig(platform operatorapi.KogitoServerlessPlatform) (map[string]
 		return nil, fmt.Errorf("unable to retrieve the namespace from platform %s", platform.Name)
 	}
 	customConfig[configKeyBuildNamespace] = platform.Namespace
-	if platform.Spec.BuildPlatform.Registry.Secret == "" {
```
The registry secret was required by Kaniko

desmax74 replied:
The registry secret was required by Kaniko

desmax74 commented on builder/config.go:
```diff
@@ -46,23 +46,22 @@ func NewCustomConfig(platform operatorapi.KogitoServerlessPlatform) (map[string]
+	// Registry Secret and Address are not required, the inner builder will use minikube inner registry if available
+	// TODO: we should review
 	customConfig[configKeyRegistrySecret] = platform.Spec.BuildPlatform.Registry.Secret
-	if platform.Spec.BuildPlatform.Registry.Address == "" {
```
The registry address was required by kaniko

ricardozanini commented:
@desmax74 thanks for your review, I remembered to open a new JIRA for prod. Please see: https://issues.redhat.com/browse/KOGITO-8806

We won't require a registry anymore. Minikube has the capability to use its internal registry. For OCP, the default setup with BuildConfig also is capable to use their internal registry. We shouldn't require these parameters. See the e2e working perfectly now with prod.

davidesalerno commented on utils/kubernetes/deployment.go:
```diff
@@ -52,12 +52,12 @@ func IsDeploymentProgressing(deployment *appsv1.Deployment) bool {
 	for _, condition := range deployment.Status.Conditions {
-		if condition.Type == appsv1.DeploymentReplicaFailure &&
-			condition.Status == v1.ConditionTrue {
+		if condition.Type == appsv1.DeploymentAvailable &&
```
Shouldn't be the deployment in status ```DeploymentUnavailableReason```

ricardozanini replied:
It might! I'll take a look at the docs again.

ricardozanini replied:
Actually, no. But I improved this function to look for two conditions in the underlying deployment. This way we can provide a more precise error message/reason to a failure.
